tensorly/decomposition/_tr.py

```diff
@@ -116,6 +116,157 @@ def tensor_ring(input_tensor, rank, mode=0, svd="truncated_svd", verbose=False):
+        i-th mode of `tensor`.
+    ls_solve : {"lstsq", "normal_eq"}, default is "lstsq"
+        When equal to "lstsq", the least squares problems are solved as overdetermined
+        systems of equations using tl.lstsq. When equal to "normal_eq", a normal
+        equation formulation is used instead together with tl.solve. This latter
+        approach is expected to yield poorer accuracy due to numerical issuse with the
+        normal equations.
+    n_iter_max : int, default is 100
+        Maximum number of ALS iterations.
+    tol : float, default 1e-6
+        The algorithm is terminated when the change in the relative reconstruction error
+        is less than `tol`.
+    random_state : {None, int, np.random.RandomState}
+        Used to set the random seed in the algorithm.
+    verbose : bool, default False
+        If True, the algorithm will make some additional print outs.
+    callback : {None, Callable[[tl.tr_tensor.TRTensor, float], {None, bool}]}, default None
+        A callback function which can be used for, e.g., logging the per-iteration
+        error. Such a function takes two inputs: A tensor ring decomposition and a float
+        which indicates the relative reconstruction error between `tensor` and the
+        inputted TRTensor. It can return a bool which can be used to control when the
+        ALS algorithm terminates.
+
+    Returns
+    -------
+    tr_decomp : tl.tr_tensor.TRTensor
+        The tensor ring decomposition computed by the algorithm.
+
+    References
+    ----------
+    .. [1] Q. Zhao, G. Zhou, S. Xie, L. Zhang, A. Cichocki, "Tensor Ring Decomposition",
+           arXiv:1606.05535, 2016.
+    """
+
+    shape = tl.shape(tensor)
+    rank = validate_tr_rank(shape, rank=rank)
+    n_dim = len(shape)
+    rng = tl.check_random_state(random_state)
+    if tol > 0 or callback:
+        tensor_norm = tl.norm(tensor)
+    valid_ls_solve = {"lstsq", "normal_eq"}
+    if ls_solve not in valid_ls_solve:
+        raise ValueError(
+            f"Invalid value provided for ls_solve. It must be in {valid_ls_solve}."
+        )
+
+    # Randomly initialize decomposition cores
+    tr_decomp = tl.random.random_tr(shape, rank, random_state=rng, **tl.context(tensor))
+
+    # Run callback function if provided
+    if callback:
+        rel_error = tl.norm(tl.tr_to_tensor(tr_decomp) - tensor) / tensor_norm
+        callback(tr_decomp, rel_error)
+
+    # Main loop
+    rec_errors = []
+    for iter in range(n_iter_max):
+        for dim in range(n_dim):
+            # Compute appropriate transposed unfolding of tensor
+            tensor_unf = tl.reshape(tl.moveaxis(tensor, dim, -1), (-1, shape[dim]))
+
+            # Compute design matrix
+            subchain_tensor = tr_decomp[(dim + 1) % n_dim]
+            for j in range(2, n_dim):
+                subchain_tensor = tl.tensordot(
+                    subchain_tensor, tr_decomp[(dim + j) % n_dim], axes=1
+                )
```

Comment:
Could we use tr_to_tensor here? The advantage is that then with the einsum backend we could find a better contraction path.

Reply:
That's a good point - perhaps we can have a more general version, with e.g. `skip_mode` and the main tr_to_tensor could just add the checks and use that.

Reply:
Thanks for this suggestion. I've been looking into if using einsum with the einsum backend (which is really nice with the built-in optimization, btw!) for the various contractions on lines 203-214 in `tensorly/decomposition/_tr.py` can speed things up. I'm ironing out the testing, but will hopefully have some results soon.

Will let you know when this is done!